### PR TITLE
BLD Uses cibuildwheel for linux + osx wheels [cd build]

### PR DESCRIPTION
**Pull request by thomasjpfan**
Related to https://github.com/MacPython/numpy-wheels/issues/116

This PR:

1. Enables x86 linux+osx wheels to be built with cibuildwheel.
2. Allows PR to trigger the wheel build `[cd build]` is in the commit message, where "cd" means "continuous delivery"

I kept this PR small by only supporting two platforms. Follow up includes:

1. Windows support
2. Cross compile universal2 support for macOS
3. Upload to nightly index at https://pypi.anaconda.org/scipy-wheels-nightly/simple
4. Maybe linux arm support using QEMU depending on how fast it is

**Note**: The big diff comes from the licenses that was copied over from numpy-wheels.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks a lot for working on this @thomasjpfan! Moving to cibuildwheel will be valuable long-term.

**Review by rgommers — COMMENTED** [review]
This looks like it will be a very pleasant change of pace, no more bash & submodules:)

**Comment by thomasjpfan** [comment]
> Uploading to the Anaconda staging bucket, and having the release manager download those and re-upload when they're ready seems safer to me both release mechanics wise and security wise

A follow up to this PR is to upload to an Anaconda staging bucket. From there you can follow your current release workflow of downloading and re-uploaded.

**Review by mattip — COMMENTED** [review]
As far as I can tell there is only one nit about a future PR to remove compilation flags, and this looks good to go. 

Edit: and the versioneer problem

**Comment by mattip** [comment]
One last thing: it would be good to document the `[cd build]` string somewhere. See also #18343 which is still open.

**Comment by mattip** [comment]
We discussed this at the weekly triage meeting. One workflow that is not clear is the desire to kick off all the different CI that will be used to build wheels against a commit (This PR uses github actions. Currently travis-ci is used for arm64, ppc64le, and s390x. We have concerns that the arm64 builds on travis are flaky and are looking for alternatives, so we may end up with a third CI service). Since we use at least two services, a button to trigger builds on github actions would not be sufficient to complete the task. Currently, the MacPython/numpy-wheels workflow allows us to make a commit to that repo, which then pulls the appropriate commit from here and builds a wheel. If we merge this PR to numpy/numpy, then we could continue with that idea: make a commit to trigger wheel building. But then that commit will become part of the permanent numpy git history. Is that the way other projects use cibuildwheel?

There is also some concern that a run of this CI job will push wheels to anaconda.org, these would be marked as the "latest dev wheels" and picked up by projects building off of latest NumPy wheels. It seems that is a bit too open. Is there a way to limit the ability to upload?

**Comment by mattip** [comment]
In recent community/triage discussions we decided to put this in. TBD:
- upload weekly to https://anaconda.org/scipy-wheels-nightly/numpy and https://anaconda.org/multibuild-wheels-staging/numpy for releases
- document the process at https://numpy.org/devdocs/dev/releasing.html

I would like to put this in as-is and do the rest in future PRs. Does that make sense?

**Comment by mattip** [comment]
Thanks @thomasjpfan

**Comment by thomasjpfan** [comment]
> I would like to put this in as-is and do the rest in future PRs. Does that make sense?

Your plan looks good to me.

File: .github/workflows/wheels.yml [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Workflow to build and test wheels.
+# To work on the wheel building infrastructure on a fork, comment out:
+#
+# if: github.repository == 'numpy/numpy'
+#
+# in the get_commit_message job. Be sure to include [cd build] in your commit
```
[the remark this answers is not in the document]
Comment (peterjc) [reply .github/workflows/wheels.yml:6]: Was there a naming rationale? I would have guessed "[wheel build]" but the meaning of "cd" here escapes me.

[the remark this answers is not in the document]
Comment (rgommers) [reply .github/workflows/wheels.yml:6]: "continuous delivery" rather than "change directory" I guess:)

[the remark this answers is not in the document]
Comment (thomasjpfan) [reply .github/workflows/wheels.yml:6]: Yea it is "continuous delivery". I think `[wheel build]` works as well.

[the remark this answers is not in the document]
Comment (rgommers) [reply .github/workflows/wheels.yml:6]: Changing in a follow up would be useful I think, `[wheel build]` is significantly more descriptive.

**Pull request merged**